packages/webamp/js/components/Vis.tsx

```diff
@@ -0,0 +1,195 @@
+  const distance = 2 * PIXEL_DENSITY;
+
+  const bgCanvasCtx = bgCanvas.getContext("2d");
+  if (bgCanvasCtx == null) {
+    throw new Error("Could not construct canvas context");
+  }
+  bgCanvasCtx.fillStyle = bgColor;
+  bgCanvasCtx.fillRect(0, 0, width, height);
+  if (!windowShade) {
+    bgCanvasCtx.fillStyle = fgColor;
+    for (let x = 0; x < width; x += distance) {
+      for (let y = PIXEL_DENSITY; y < height; y += distance) {
+        bgCanvasCtx.fillRect(x, y, PIXEL_DENSITY, PIXEL_DENSITY);
+      }
+    }
+  }
+  return bgCanvas;
+}
+
+export default function Vis({ analyser }: Props) {
+  useLayoutEffect(() => {
+    analyser.fftSize = 1024;
+  }, [analyser, analyser.fftSize]);
+  const colors = useTypedSelector(Selectors.getSkinColors);
+  const mode = useTypedSelector(Selectors.getVisualizerStyle);
+  const audioStatus = useTypedSelector(Selectors.getMediaStatus);
+  const getWindowShade = useTypedSelector(Selectors.getWindowShade);
+  const getWindowOpen = useTypedSelector(Selectors.getWindowOpen);
+  isMWOpen = getWindowOpen("main");
+  doubled = useTypedSelector(Selectors.getDoubled);
+  const dummyVizData = useTypedSelector(Selectors.getDummyVizData);
+  const toggleVisualizerStyle = useActionCreator(Actions.toggleVisualizerStyle);
+  const windowShade = getWindowShade("main");
+
+  smallVis = windowShade && isMWOpen;
+  const renderWidth = 76;
+  const renderWidthBG = !isMWOpen 
+  ? renderWidth 
+  : windowShade 
+    ? (doubled ? renderWidth : 38)
+    : renderWidth * PIXEL_DENSITY;
+
+  renderHeight = smallVis ? 5 : 16;
+  PIXEL_DENSITY = doubled && smallVis ? 2 : 1;
+
+  const width = renderWidth * PIXEL_DENSITY;
+  const height = renderHeight * PIXEL_DENSITY;
+
+  const bgCanvas = useMemo(() => {
+    return preRenderBg(
+      renderWidthBG,
+      height,
+      colors[0],
+      colors[1],
+      Boolean(windowShade)
+    );
+  }, [colors, height, width, windowShade]);
+
+  const [canvas, setCanvas] = useState<HTMLCanvasElement | null>(null);
+
+  //? painter administration
+  const [painter, setPainter] = useState<VisPaintHandler | null>(null);
+
+  useEffect(() => {
+    if (!canvas) return;
+    const _setPainter = (PainterType: typeof VisPaintHandler) => {
+      const _vis: IVis = {
+        canvas,
+        colors,
+        analyser,
+        oscStyle: "lines",
+        bandwidth: "wide",
+        coloring: "normal",
+        peaks: true,
+        saFalloff: "moderate",
+        saPeakFalloff: "slow",
+        sa: "analyzer",
+      };
+      const newPainter = new PainterType(_vis);
+      setPainter(newPainter);
+    };
+    switch (mode) {
+      case VISUALIZERS.OSCILLOSCOPE:
+        _setPainter(WavePaintHandler);
+        break;
+      case VISUALIZERS.BAR:
+        _setPainter(BarPaintHandler);
+        break;
+      case VISUALIZERS.NONE:
+        _setPainter(NoVisualizerHandler);
+        break;
+      default:
+        _setPainter(NoVisualizerHandler);
+    }
+  }, [analyser, canvas, mode, colors]);
```

Comment:
I think this could be a `useMemo`
```suggestion
  const painter, setPainter = useMemo(() => {
    if (!canvas) return null;
    const vis: IVis = {
        canvas,
        colors,
        analyser,
        oscStyle: "lines",
        bandwidth: "wide",
        coloring: "normal",
        peaks: true,
        saFalloff: "moderate",
        saPeakFalloff: "slow",
        sa: "analyzer",
     };
    switch (mode) {
      case VISUALIZERS.OSCILLOSCOPE:
        return new WavePaintHandler(vis);
      case VISUALIZERS.BAR:
        return new BarPaintHandler(vis);
      case VISUALIZERS.NONE:
        return new NoVisualizerHandler(vis);
      default:
        return new NoVisualizerHandler(vis);
    }
  }, [analyser, canvas, mode, colors]);
```